examples/webgl_instancing_modified.html

```diff
@@ -0,0 +1,196 @@
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<title>three.js webgl - instancing - modified</title>
+		<meta charset="utf-8">
+		<meta name="viewport" content="width=device-width, user-scalable=no, minimum-scale=1.0, maximum-scale=1.0">
+		<link type="text/css" rel="stylesheet" href="main.css">
+	</head>
+
+	<body>
+		<div id="info">
+			<a href="http://threejs.org" target="_blank" rel="noopener">three.js</a> - webgl instancing - modified
+		</div>
+
+		<script type="module">
+
+			import * as THREE from '../build/three.module.js';
+
+			import Stats from './jsm/libs/stats.module.js';
+
+			var camera, scene, renderer, stats;
+
+			var mesh;
+
+			var dummy = new THREE.Object3D();
+
+			var amount = 8;
+			var count = Math.pow( amount, 3 );
+
+			init();
+			animate();
+
+			function init() {
+
+				renderer = new THREE.WebGLRenderer( { antialias: false } ); // false improves the frame rate
+				renderer.setPixelRatio( window.devicePixelRatio );
+				renderer.setSize( window.innerWidth, window.innerHeight );
+				document.body.appendChild( renderer.domElement );
+
+				renderer.outputEncoding = THREE.sRGBEncoding;
+
+				scene = new THREE.Scene();
+
+				camera = new THREE.PerspectiveCamera( 40, window.innerWidth / window.innerHeight, 1, 100 );
+				camera.position.set( 0, 0, 20 );
+
+				new THREE.BufferGeometryLoader().load( 'models/json/suzanne_buffergeometry.json', function ( geometry ) {
+
+					var instanceColors = [];
+
+					for ( var i = 0; i < count; i ++ ) {
+
+						instanceColors.push( 0.5 + 0.5 * Math.random() );
+						instanceColors.push( 0.5 + 0.5 * Math.random() );
+						instanceColors.push( 0.5 + 0.5 * Math.random() );
+
+					}
+
+					geometry.setAttribute( 'instanceColor', new THREE.InstancedBufferAttribute( new Float32Array( instanceColors ), 3 ) );
+
+					geometry.computeVertexNormals();
+
+					geometry.scale( 0.5, 0.5, 0.5 );
+
+					//console.log( geometry );
+
+					//
+
+					new THREE.TextureLoader().load( 'textures/matcaps/matcap-porcelain-white.jpg', function ( texture ) {
+
+						texture.encoding = THREE.sRGBEncoding;
+
+						var material = new THREE.MeshMatcapMaterial( { color: 0xaaaaff, matcap: texture } );
+
+						var colorParsChunk = [
+							'attribute vec3 instanceColor;',
+							'varying vec3 vInstanceColor;',
+							'#include <common>'
+						].join( '\n' );
+
+						var instanceColorChunk = [
+							'#include <begin_vertex>',
+							'\tvInstanceColor = instanceColor;'
+						].join( '\n' );
+
+						var fragmentParsChunk = [
+							'varying vec3 vInstanceColor;',
+							'#include <common>'
+						].join( '\n' );
+
+						var colorChunk = [
+							'vec4 diffuseColor = vec4( diffuse * vInstanceColor, opacity );'
+						].join( '\n' );
+
```

Comment:
How much trouble would it be to make this modification offset the UVs rather than changing the color — aside from the effort of finding a model and texture that look good with some offsets? 🤔

There's an easier way to set per-instance color, which I included in `instancing / scatter`:

https://github.com/mrdoob/three.js/blob/45418089bd5633e856384a8c0beefced87143334/examples/webgl_instancing_scatter.html#L82-L98

^But if for some reason that's not a recommended way to do this, I'm happy to address it there.

Reply:
I'm imagining a translation-only UV offset... so e.g. four sub-textures could be packed adjacent in one image...

. 1 – 2 – 3 – 4
◾️◾️◾️◾️

... and the offset would select a different sub-texture for different instances. That would seem like a good example for something like minecraft-style voxels, where there are a finite number of voxel types.

Reply:
```js
blossomGeometry.setAttribute( 'color', new THREE.InstancedBufferAttribute( color, 3 ) );

blossomMesh = new THREE.InstancedMesh( blossomGeometry, blossomMaterial, count );
```

Yikes. I can't imagine adding an  `InstancedBufferAttribute` to `BufferGeometry` ever working. I would not recommend we encourage that coding pattern -- even if it does work in this case.

>^But if for some reason that's not a recommended way to do this, I'm happy to address it there.

Yes, I'd recommend you think about changing that.

Reply:
It's `InstancedMesh` + `InstancedBufferGeometry` + `InstancedBufferAttribute` there, which doesn't seem obviously wrong, but ¯\\_(ツ)_/¯

Reply:
It seems to me that `InstancedBufferAttribute` should only be added to `InstancedBufferGeometry`. Adding it to regular `BufferGeometry` is inappropriate -- IMO.

Reply:
There is no `BufferGeometry` in any of these examples — I'm passing `InstantedBufferGeometry` to the `InstancedMesh` constructor.

Reply:
> I'm passing InstantedBufferGeometry to the InstancedMesh constructor.
```js
blossomGeometry.setAttribute( 'color', new THREE.InstancedBufferAttribute( color, 3 ) );
```
Is not `blossomGeometry` is an instance of `BufferGeometry`?

>I'm passing `InstancedBufferGeometry` to the `InstancedMesh` constructor.

The `InstancedMesh` constructor expects a `BufferGeometry`, no?

Reply:
Ahh you're right. In my head that was definitely InstancedBufferGeometry. 🤔

I would expect the `InstancedMesh` constructor to accept `InstancedBufferGeometry`, to allow users to keep its API even when using other instanced attributes, but apparently I haven't actually tested that yet!

Reply:
^But I was not expecting BufferGeometry+InstancedBufferAttribute to work. I'll update that example...